genesis/src/lib.rs

```diff
@@ -34,58 +33,66 @@ pub struct GenesisTransaction(pub SignedTransaction);
 /// [`GenesisNetwork`] contains initial transactions and genesis setup related parameters.
 #[derive(Debug, Clone)]
 pub struct GenesisNetwork {
-    /// transactions from `GenesisBlock`, any transacton is accepted
-    pub transactions: Vec<GenesisTransaction>,
+    /// Transactions from [`RawGenesisBlock`]. This vector is guaranteed to be non-empty,
+    /// unless [`GenesisNetwork::transactions_mut()`] is used.
+    transactions: Vec<GenesisTransaction>,
 }
 
 impl GenesisNetwork {
     /// Construct [`GenesisNetwork`] from configuration.
     ///
     /// # Errors
-    /// Fails if genesis block is not found or cannot be deserialized.
-    pub fn from_configuration(
-        raw_block: RawGenesisBlock,
-        genesis_config: Option<&Configuration>,
-    ) -> Result<GenesisNetwork> {
-        iroha_logger::debug!("Submitting genesis.");
-        let genesis_config =
-            genesis_config.expect("Should be `Some` when `submit_genesis` is true");
-        let genesis_key_pair = KeyPair::new(
-            genesis_config.account_public_key.clone(),
-            genesis_config
-                .account_private_key
-                .clone()
-                .ok_or_else(|| eyre!("Genesis account private key is empty."))?,
-        )?;
+    /// - If fails to sign a transaction (which means that the `key_pair` is malformed rather
+    ///   than anything else)
+    /// - If transactions set is empty
+    pub fn new(raw_block: RawGenesisBlock, genesis_key_pair: &KeyPair) -> Result<GenesisNetwork> {
         #[cfg(not(test))]
         // First instruction should be Executor upgrade.
         // This makes possible to grant permissions to users in genesis.
         let transactions_iter = std::iter::once(GenesisTransactionBuilder {
-            isi: vec![Upgrade::new(Executor::try_from(raw_block.executor)?).into()],
+            isi: vec![Upgrade::new(
+                Executor::try_from(raw_block.executor)
+                    .wrap_err("Failed to construct the executor")?,
+            )
+            .into()],
         })
         .chain(raw_block.transactions);
 
         #[cfg(test)]
-        let transactions_iter = raw_block.transactions.into_iter();
+        let transactions_iter = {
+            if raw_block.transactions.is_empty() {
+                eyre::bail!("Genesis transaction set is empty");
```

Comment:
Hm, this doesn't error out if `iroha_genesis` is not compiled for unit tests. Is this the logic you intended?

(it's weird that we have different behavior with `cfg(test)`.. Do you know why is that?)

Reply:
I found some weird code in this function:

- In `cfg(test)` it doesn't put Executor in the transactions
- Then it signs each transaction and filters out failing ones, although if one fails, every other one will also fail
- Then, it checks if the resulting transactions set is empty and `bail`s if so

I made a conclusion that it is possible to get empty transactions set only in `cfg(test)`. 

Anyway, I think it is better to remove this logic out of here completely to the `TestGenesis` you mentioned in another discussion.

Reply:
Ah, and more context:

- https://github.com/hyperledger/iroha/issues/3075